src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/SegmentedList`1.cs

```diff
@@ -0,0 +1,794 @@
+        {
+            return BinarySearch(in item, 0, _count - 1, comparer);
+        }
+
+        /// <summary>
+        /// Performs a binary search in a sorted list.
+        /// </summary>
+        /// <param name="item">Element to search for.</param>
+        /// <param name="low">The lowest index in which to search.</param>
+        /// <param name="high">The highest index in which to search.</param>
+        /// <param name="comparer">Comparer to use.</param>
+        /// <returns>The index </returns>
+        public int BinarySearch<TComparer>(in T item, int low, int high, TComparer comparer)
+            where TComparer : IByRefComparer<T>
+        {
+            if (low < 0 || low > high)
+            {
+                throw new ArgumentOutOfRangeException($"Low index, with value {low}, must not be negative and cannot be greater than the high index, whose value is {high}.");
+            }
+
+            if (high < 0 || high >= _count)
+            {
+                throw new ArgumentOutOfRangeException($"High index, with value {high}, must not be negative and cannot be greater than the number of elements contained in the list, which is {_count}.");
+            }
+
+            while (low <= high)
+            {
+                var i = low + ((high - low) >> 1);
+                var order = comparer.Compare(in _items[i >> _segmentShift][i & _offsetMask], in item);
+
+                if (order == 0)
+                {
+                    return i;
+                }
+
+                if (order < 0)
+                {
+                    low = i + 1;
+                }
+                else
+                {
+                    high = i - 1;
+                }
+            }
+
+            return ~low;
+        }
+
+        /// <summary>
+        /// Sorts the list using default comparer for elements.
+        /// </summary>
+        public void Sort()
+        {
+            Sort(Comparer<T>.Default);
+        }
+
+        /// <summary>
+        /// Sorts the list using specified comparer for elements.
+        /// </summary>
+        /// <param name="comparer">Comparer to use.</param>
+        public void Sort(IComparer<T> comparer)
+        {
+            if (_count <= 1)
+            {
+                return;
+            }
+
+            QuickSort(0, _count - 1, comparer);
+        }
+
+        /// <summary>
+        /// Appends a range of elements from anothe list.
+        /// </summary>
+        /// <param name="from">Source list.</param>
+        /// <param name="index">Start index in the source list.</param>
+        /// <param name="count">Count of elements from the source list to append.</param>
+        public void AppendFrom(SegmentedList<T> from, int index, int count)
+        {
+            if (count > 0)
+            {
+                var minCapacity = _count + count;
+
+                if (_capacity < minCapacity)
+                {
+                    EnsureCapacity(minCapacity);
```

Comment:
feels like we could just have callers call EnsureCapacity, and EnsureCapacity can early bail if necessary.